Final Development

knightAmor97 opened the pull request:
Added the extra features and commented through the code

knightAmor97 committed:
Final changes

I've added the extra requirement and commented the code more

knightAmor97 committed:
program.cs changes

Joshdunny commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,69 +1,178 @@
 
-        //initaites these in the constructor
-        public Game()
+        //polymorphism is used in order to change the values but they cannot be edited outside the program
```
I'm not sure polymorphism is the right phrase you are looking for, encapsulation is the correct term for your description here

knightAmor97 replied:
That's been changed just a little mistake

Joshdunny commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,69 +1,178 @@
         /*
-         * sumup methods will roll all 3 dice 
-         * it will then add the dice together and return the result
+         * these are the methods for this class
+         * SumUp should addup all 3 die and then store the results inside the the list it will then return it
+         * amountOfRolls allows the user to reroll as many times as the user wants
+         * reRoll rerolls the die from the amount the user has given
+         * addUpAll should add up all the dies roll together
+         * averageDie will take all the die added up together then divide it by the amount the user has asked to reroll
+         * minResults will find the minimum score of what the user got
+         * maxResults will find the max score the user got
+         * displayInformation will display all the data gathered
```
These are some great comments i would suggest putting them above their respective methods instead of a large chunk here, however this is a personal opinion so it is up to you.

knightAmor97 replied:
that was just for me so I know  what to do but its been changed

Joshdunny commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,69 +1,178 @@
+                //adds all the results it gathered together
+                SumOfAll += results[i];
+            }
+            //returns the results together
+            return SumOfAll;
+        }
+        public double averageDie()
+        {
+            //sets up a variable and assigns it to the addUpall method
+            int sumOfAll = AddUpAll();
+            //gets the total amount of die and dividing it by the amount of die there is (gotten by taking the noOfRolls x3) to get the average
+            double average = (double)sumOfAll / (noOfRolls *3);
+            //returns a rounded result (rounded by 2 dp)
+            return Math.Round(average, 2);
         }
 
+        public int minResult()
+        {
+            //finds the minimum result that the user got
+            int minResult = results.Min();
+            //returns the minimum result
+            return minResult;
+        }
 
+        public int maxResult()
+        {
+            //finds the maximum result that the user got
+            int maxResult = results.Max();
+            //returns the result
+            return maxResult;
+        }
 
+        public void displayInformation()
+        {
+            //outputs the average, min, max scores and the sum up all the scores the user got
+            Console.WriteLine($" sum of all the die: {AddUpAll()}");
+            Console.WriteLine($"average score the user got: {averageDie()}");
+            Console.WriteLine($"minimum score you got: {minResult()}");
+            Console.WriteLine($"maximum score the user got: {maxResult()}");
+        }
```
This is a great way of breaking up the Game class into different methods

knightAmor97 replied:
thank you

Joshdunny commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,69 +1,178 @@
-            int D1Roll = _D1.roll();
-            int D2Roll = _D2.roll();
-            int D3Roll = _D3.roll();
+            int D1Roll = D1.roll();
+            int D2Roll = D2.roll();
+            int D3Roll = D3.roll();
```
In my opinion you could save some memory you don't need to assign the rolled number to a variable and instead access the value using the objects methods when you are accessing the value.

knightAmor97 replied:
Changed it so now that its now assigned within the class and then rolls within the method

Joshdunny commented on CMP1903_A1_2324/Die.cs:
Your Die class looks well done, it has all the requirements.

You have formatted your code using almost all of the style guide besides a few things like using tab instead of 2 spaces.

Very easy to read especially with the comments 

No errors occur during the use of the class and its methods

knightAmor97 replied:
Thank you the style changes should've been changed although it might just be a GitHub/browser thing to fit it in

Joshdunny commented on CMP1903_A1_2324/Game.cs:
The Game class meets all of the requirements including the Extra ones which are implemented

Formatting again almost followed all the way using the style guide

commenting is done nicely with just one personal suggestion about the commenting, therefore very readable code

However errors are occuring when ambiguous input is given to AmountOfRolls

knightAmor97 replied:
I've added code to stop any erroneous inputs from happening, does anything else need changing

Joshdunny replied:
Could you push the changes to github so i can see

Joshdunny commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,69 +1,178 @@
-        //These methods will get and return all 3 of the dice objectsbut cannot be edited as it is a private object
-        public Die d1
+        public void AmountOfRolls()
         {
-            get { return _D1; }
+            //asks how much the user wants to roll then takes the users input
+            Console.WriteLine("How many times would you like to reroll");
+            int numberOfRolls = int.Parse(Console.ReadLine());
+            // will set the built in variable to what the user has inputted
+            noOfRolls = numberOfRolls;
```
Error not handled correctly, when ambiguous input is given an error occurs causing the program to crash

Joshdunny commented on CMP1903_A1_2324/Program.cs:
```diff
@@ -27,13 +26,40 @@ static void Main(string[] args)
-            Console.ReadKey();
+            bool EnteredCorrectly = false;
             
+            // do loop used so the user doesnt input something else
+            do
+            {
+                //asks user if they want to reroll
+                Console.WriteLine("would you like to reroll(y/n)");
+                string answer = Console.ReadLine();
+                // if they answered yes it will go thrugh the amountOfRolls, reroll and then it will display all the statistics
+                if (answer == "y")
+                {
+                    game.AmountOfRolls();
+                    game.reroll();
+                    game.displayInformation();
+                    //this will exit the loop
+                    EnteredCorrectly = true;
+                }
+                // if they dont want to reroll it exit the loop
+                else if (answer == "n")
+                {
+                    EnteredCorrectly = true;
+                }
+                // if the user didnt enter a correct answer it will ask them to input again
+                else
+                {
+                    Console.WriteLine("didnt enter a right input please try again");
+                }
+            } while (EnteredCorrectly == false);
+            Console.ReadKey();
```
Good way of handling the users input to get a yes or no

Joshdunny commented on CMP1903_A1_2324/Program.cs:
Program has all requirements 

Style guide followed to a point

Comments are well done for readable code

Errors with user input in this class done well

Joshdunny commented on CMP1903_A1_2324/Testing.cs:
Testing is well done with great use of debug.assert(), however when testing the sum, sum does not seem to be tested for if the sum is out of the possible range and instead testing the method is correctly summing values. If this is the intended implementation that is fine.

Otherwise all requirements are met 

Style guide is followed mostly besides indentation  

Errors are handled correctly

knightAmor97 replied:
Changed it so now it checks the range of the score as well as checking if its adding up correctly

Joshdunny reviewed (commented):
All in All this code is very well done! All requirements are added even the extra ones.

Formatting is mostly followed besides indentation, if you would like to adhere to the style guide i would implement this

Commenting is done throughout and done nicely for easily readable code

Not all errors are handled correctly with ambiguous input, i have noted out where the error occurs other then that error no more errors occur and input in other places is handled nicely

knightAmor97 committed:
changes from code review #1

WonderboyMattT commented on CMP1903_A1_2324/Die.cs:
The structure is good, with encapsulation and separation for easy reading.
Consider changing some of your naming conventions, such as "rand" for something more descriptive like _random, and value to CurrentValue. Similarly, your roll() method should be capitalised.

Functionally, moving the output on the roll() function to the calling code is compliant with C# conventions. Also, remove the "value = 0" as setting your Die constructor to 0 could cause confusion since the given die range is 1-6.

knightAmor97 replied:
(don't need to respond to the comments) I've changed it so that value is not set to 0

WonderboyMattT commented on CMP1903_A1_2324/Game.cs:
Your constructor initialises 3 die objects along with a list to store the result. Consider renaming them Dice1, Dice2 etc instead of D* for ease of reading.
I like the reroll() and AmountOfRolls() functionality and how the methods behave consistently no matter the amount of rerolls desired. I especially like the use of the try-catch block on  the latter to handle incorrect inputs. 

Line 245 "average die the you got" 

When launching the game, there is a pause of a few seconds after the initial user prompt and I can't identify what's causing this. I think it might be caused by the series of getters and setters in the Game class slowing down the processing. For example, SumOfAll doesn't need a value setting.

Furthermore, a lot of the code repeats itself, whereas you could use Game to construct a list, then call the Die class to instantiate a new Die object when required. This would dramatically shorten the amount of code an potentially mitigate the pause at the start of the game.

knightAmor97 replied:
Changed the code so that there aren't as many getters and setters and changed the die name so that its now dice1 instead of d1 etc

WonderboyMattT commented on CMP1903_A1_2324/Testing.cs:
I'm not sure what Console.Clear() does as I removed it during testing an increased the range on the die to 1-9. Without the clear, the result remains on the console, which would be better for identifying errors alongside the error message you already have.
Otherwise, the structure is good and you use Debug.Assert as required. Add a comment to explain what each Debug.Assert is for.

knightAmor97 replied:
Removed the console.clear() and instead labelled the testing phase in main so it shows the user the die is rolling in a range from 1-6 then clears the console when there happy with to avoid confusion with the actual die they are rolling.

WonderboyMattT commented on CMP1903_A1_2324/Program.cs:
You have really clear prompts for the user, which I always appreciate, and the outputs are clear and concise too, further enhancing the experience. 
Using do-while loop allows the user to try again after an incorrect input, which is very considerate. You seem very aware of the fallibility of human error.

WonderboyMattT reviewed (commented):
It's really good and runs well from the users perspective, with some appealing features. I like how well you have implemented the extra task and how consistently it performs. You've easily met the criteria of the assignment.

Some of the naming conventions needs updating to be consistent with the style guide.
Comments are good, though some could be altered to be more concise in how they describe what the code is doing.

knightAmor97 committed:
changes for code review #2

Removed value set to 0, Changed the names of some variables and methods to match more closely to the style guide, removed some getters and setters to make the code faster and changed how the testing phase works so that the user can see that the die is rolling correctly.
